spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelCommandDynamicPruning.scala

```diff
@@ -86,6 +89,16 @@ case class RowLevelCommandDynamicPruning(spark: SparkSession) extends Rule[Logic
     val matchingRowsPlan = command match {
       case d: DeleteFromIcebergTable =>
         Filter(d.condition.get, relation)
+      case u: UpdateIcebergTable =>
+        // UPDATEs with subqueries may be rewritten using a UNION with two identical scan relations
+        // each scan relation will get its own dynamic filter that will be shared during execution
+        // the analyzer will assign different expr IDs for each scan relation output attributes
+        // that's why the condition may refer to invalid attr expr IDs and must be transformed
```

Comment:
If I understand correctly, sometimes a plan gets transformed, from this:

```
ReplaceData
  Update(id#1 IN (1, 5), data#2 = 'foo')
      V2Relation(db.table, [id#1, data#2])
```

To this:
```
ReplaceData
  Update(id#1 IN (1, 5), data#2 = 'foo')
    Union([id#1, data#2])
      V2Relation(db.table, [id#1, data#2])
      V2Relation(db.table, [id#4, data#5])
```

So this is basically creating the dynamic filter for each scan separately and fixing up the attrs, from the update's IDs to the relation's IDs.

Is there a test case for this, or is it something that you ran into in practice?

It took me a while to understand this (assuming that I do) and I think it would be nice if it were more clear what is being replaced. Renaming `attrMap` to `scanRelationAttrs` would probably help so it is clear we're looking up the right attr for the scan relation.

It would also be good to to add a comment that explains why `u.table.output` is always going to match `relation.output`. In fact, I'm not sure that will. What happens when a column in the data is not used to produce the output? Is it pruned?

For example, `UPDATE t SET id = 1 WHERE true` doesn't need to project `id` from the table. If `id` is not projected, then the `attrMap` is incorrect. Is there a reason why that can't happen?

Reply:
Okay, after reading through all the cases for rewriting update, this makes a lot more sense. I think it would have helped me reading through this to either direct to the union case in `RewriteUpdateTable` or to explain the rewrite here and give a little more context on when there will be a union.

As for my other comments, I think it would still be good to rename `attrMap` and I'm not sure whether there could be a case where `u.table.output` doesn't match `relation.output`.

Reply:
I've tried to improve the comment to make it clear and also made a note about the rewrite rule. Let me know if its any better.

> For example, UPDATE t SET id = 1 WHERE true doesn't need to project id from the table. If id is not projected, then the attrMap is incorrect.

It was working correctly as we don't support schema pruning in copy-on-write operations. The only case where it is applicable if we have `true` update condition and then we can discard only columns assigned to literals. Seemed like a very narrow use case. I think we can handle it later as we did not support it in the old implementation. If the condition is not trivial, the assigned expression will be `if ... else oldValue` so we won't be able to prune the column.

That being said, I did modify the logic for building the attribute map. It should work even if we add schema pruning.